### PR TITLE
basic message filtering

### PR DESCRIPTION
**Pull request by dmacthedestroyer**
* Publishers can attach metadata to their messages in the form of key:value pairs
* Subscribers can apply a filter to their queues in the form of key:value pairs
* In order for a subscriber to receive a message with a filter, every key:value pair condition be a match
* Matching is case-sensitive exact string matching

Current Limitations
* can't delete an existing message filter due to shortcomings in the AWS API
* can only do exact string matching 
** AWS API supports int ranges, string startswith, array contains, and blacklisting operations as well, but I didn't include these because I'm not sure what kind of capabilities other broker implementations can support. This can be expanded if desired.

### CODE REVIEW

File: aws/subscriber.go [hunk 1]
```diff
@@ -119,9 +96,101 @@ func (m *awsMessage) MessageID() string {
...
+	s.wg.Add(1)
+	defer s.wg.Done()
+	resp, err := s.sqs.ReceiveMessageWithContext(ctx, &sqs.ReceiveMessageInput{
+		QueueUrl:            s.queueURL,
+		WaitTimeSeconds:     aws.Int64(20),
+		MaxNumberOfMessages: aws.Int64(1),
+	})
+	if err != nil {
+		errChannel <- err
+	} else {
```
Comment (johnbelamaric) [inline aws/subscriber.go]: Just return here then you don't need the else.

File: aws/subscriber.go [hunk 2]
```diff
@@ -119,9 +96,101 @@ func (m *awsMessage) MessageID() string {
...
+		for _, msg := range resp.Messages {
+			message, err := decodeFromSQSMessage(msg.Body)
+			metadata := decodeMessageAttributes(msg.MessageAttributes)
+			if err != nil {
+				errChannel <- err
+			} else {
```
Comment (johnbelamaric) [inline aws/subscriber.go]: Just `continue` here and you can get rid of the `else`.

File: aws/aws.go [hunk 3]
```diff
@@ -193,31 +153,60 @@ func ensureQueuePolicy(queueURL, topicArn *string, sqsClient sqsiface.SQSAPI) er
...
-	return aws.String(base64.StdEncoding.EncodeToString(msg)), nil
+func encodeToSNSMessage(msg []byte) *string {
+	return aws.String(base64.StdEncoding.EncodeToString(msg))
+}
+
+// encodeMessageAttributes converts the given metadata map into input needed for
+// sns.PublishInput.MessageAttributes
+func encodeMessageAttributes(metadata map[string]string) map[string]*sns.MessageAttributeValue {
+	attributes := make(map[string]*sns.MessageAttributeValue)
+	if metadata != nil {
```
Comment (johnbelamaric) [inline aws/aws.go]:
```
if metadata == nil {
 return attributes
}
```

Then you don't need to indent the next bit which is a bit easier to read.

File: aws/aws_test.go [hunk 4]
```diff
@@ -114,77 +117,138 @@ func TestEnsureQueuePolicy(t *testing.T) {
...
+// 		}
+//
+// 		actualFilter := *sqsSpy.spiedSetQueueAttributesInput.Attributes["FilterPolicy"]
+// 		expectedFilter, _ := encodeFilterPolicy(filter)
+//
+// 		if expectedFilter != actualFilter {
+// 			t.Errorf("expected filter to be %q, but was %q", expectedFilter, actualFilter)
+// 		}
+// 	}
+// }
```
Comment (johnbelamaric) [inline aws/aws_test.go]: Do you plan to resurrect this? If not, just delete it all. FYI /* */ is easier for block comments.

File: aws/aws_test.go [hunk 5]
```diff
@@ -114,77 +117,138 @@ func TestEnsureQueuePolicy(t *testing.T) {
...
+
+func TestEncodeFilterPolicy(t *testing.T) {
+	{
```
Comment (johnbelamaric) [inline aws/aws_test.go]: What's the purpose of putting this in a block?

Comment (dmacthedestroyer) [reply aws/aws_test.go]: nothing -- just sloppy code. Will fix

Comment (dmacthedestroyer) [reply aws/aws_test.go]: none -- just sloppy code. Will fix

**Comment by dmacthedestroyer** [comment]
I found the pattern with the blocks while reading through the go-kit repository, and thought it helped with readability -- especially regarding error checking. It creeps me out to reuse an `err` local variable, and it makes the short variable declaration clause messy. Putting them in blocks ensures their scope is limited to the particular "errorable" thing that's happening.

I'm not sure how popular it is, but in app code like `ensureSubscription` it helps me reason about each of the individual "errorable" steps required to build a subscription.

For testing, I haven't found many good examples on the preferred structure to writing tests but it doesn't seem like the traditional way of having a single function test a single aspect is the go idiom.

I'm still trying to grok Go, so please let me know if I'm developing any antipatterns.

I'll address your other comments with another commit very soon.

**Comment by dmacthedestroyer** [comment]
The AWS API uses `*string` pretty much everywhere... I just forwarded the types they use where appropriate instead of un-pointing and re-pointing between calls.

This shouldn't be a thing outside of the `aws` package, and it's not a thing for the exported functions within `aws`, either.

**Comment by dmacthedestroyer** [comment]
@johnbelamaric what else do you think is needed for this PR to go through?

**Comment by johnbelamaric** [comment]
I haven't seen that pattern of blocks in other go code. Not sure if it's considered idiomatic or not. Probably something to look into before going wild with it.

**Pull request merged**